giskard/llm/talk/tools/predict.py

```diff
@@ -0,0 +1,153 @@
+    ----------
+    default_name : str
+        The default name of the Tool. Can be re-defined with constructor.
+    default_description: str
+        The default description of the Tool's functioning. Can be re-defined with constructor.
+    """
+
+    default_name: str = "predict"
+    default_description: str = ToolDescription.PREDICT.value
+
+    @property
+    def specification(self) -> str:
+        """Return the Tool's specification in a JSON Schema format.
+
+        Returns
+        -------
+        str
+            The Tool's specification.
+        """
+        feature_json_type = get_feature_json_type(self._dataset)
+
+        return {
+            "type": "function",
+            "function": {
+                "name": self.name,
+                "description": self.description,
+                "parameters": {
+                    "type": "object",
+                    "properties": {
+                        "features_dict": {
+                            "type": "object",
+                            "properties": {
+                                feature: {"type": dtype} for feature, dtype in list(feature_json_type.items())
+                            },
+                        }
+                    },
+                    "required": ["features_dict"],
+                },
+            },
+        }
+
+    def _get_input_from_dataset(self, row_filter: dict) -> pd.DataFrame:
+        """Get input from dataset.
+
+        Filter rows from the dataset, using the `row_filter`.
+
+        Parameters
+        ----------
+        row_filter : dict
+            The dictionary with features and related values to filter the dataset.
+
+        Returns
+        -------
+        pd.DataFrame
+            The DataFrame with filtered rows.
+        """
+        threshold = 85
+        filtered_df = self._dataset.df.copy()
+
+        for col_name, col_value in list(row_filter.items()):
+            if filtered_df[col_name].dtype == "object":
+                filtered_df = filtered_df[
+                    filtered_df[col_name].apply(lambda x: fuzz.ratio(x.lower(), col_value.lower()) >= threshold)
+                ]
+                if not len(filtered_df):
+                    break
+            else:
+                filtered_df = filtered_df[filtered_df[col_name] == col_value]
+
+        return filtered_df
+
+    def _get_input_from_user(self, feature_values: dict) -> pd.DataFrame:
+        """Form input based on the user's query.
+
+        Take input feature values from user query. For the missed values, take the feature mean from the Dataset.
+
+        Parameters
+        ----------
+        feature_values : dict
+            The dictionary with features and related values extracted from the query.
+
+        Returns
+        -------
+        pd.DataFrame
+            The DataFrame with the formed input.
+        """
+        # Prepare background sample.
+        from giskard.models.model_explanation import _get_background_example
```

Comment:
1) I wanted to use the same code for the "background sample calculation" from the `giskard.models.model_explanation` module, without copying it;
2) If I put this import into top level, I am getting the circular import error.

These requirements brought the side-effect of a necessity of non-top level hidden function import.

Reply:
Could you give me the circular import you are getting ? Also, can you document it (for every import causing a circular one) ?

Reply:
Leave it as it is. Commented placed with the non-top-level import with the reason of doing this:
```
# Import the private function from the 'model_explanation' module to avoid code duplication.
# Also, if import from the top-level, we get the next error: "ImportError: cannot import name 'BaseModel'
# from partially initialized module 'giskard.models.base' (most likely due to a circular import)"
```